### PR TITLE
VideoIO classes refactoring

### PR DESCRIPTION
**Pull request by VadimLevin**
Changes:
- Added `explicit` to `VideoCapture` constructors with 2
    arguments, 1 of them has default value
- Applied library code style
- Introduced 2 debug macros to improve readability of the code

### CODE REVIEW

**Comment by VadimLevin** [comment]
This change breaks ABI compatibility. @alalek what do you think about it?

File: modules/videoio/include/opencv2/videoio.hpp [hunk 1]
```diff
@@ -628,7 +628,7 @@ class CV_EXPORTS_W VideoCapture
     implementation if multiple are available: e.g. cv::CAP_FFMPEG or cv::CAP_IMAGES or cv::CAP_DSHOW.
     @sa The list of supported API backends cv::VideoCaptureAPIs
     */
-    CV_WRAP VideoCapture(const String& filename, int apiPreference = CAP_ANY);
+    CV_WRAP explicit VideoCapture(const String& filename, int apiPreference = CAP_ANY);
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: this is useful change

File: modules/videoio/include/opencv2/videoio.hpp [hunk 2]
```diff
@@ -640,13 +640,13 @@ class CV_EXPORTS_W VideoCapture
 
     @sa The list of supported API backends cv::VideoCaptureAPIs
     */
-    CV_WRAP VideoCapture(int index, int apiPreference = CAP_ANY);
+    CV_WRAP explicit VideoCapture(int index, int apiPreference = CAP_ANY);
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: useful

File: modules/videoio/include/opencv2/videoio.hpp [hunk 3]
```diff
@@ -640,13 +640,13 @@ class CV_EXPORTS_W VideoCapture
...
 
     /** @brief Default destructor
 
     The method first calls VideoCapture::release to close the already opened file or camera.
     */
-    virtual ~VideoCapture();
+    ~VideoCapture();
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: no; please, undo it

File: modules/videoio/include/opencv2/videoio.hpp [hunk 4]
```diff
@@ -657,7 +657,7 @@ class CV_EXPORTS_W VideoCapture
 
     The method first calls VideoCapture::release to close the already opened file or camera.
      */
-    CV_WRAP virtual bool open(const String& filename, int apiPreference = CAP_ANY);
+    CV_WRAP bool open(const String& filename, int apiPreference = CAP_ANY);
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: please, restore all the virtual. no reason to change it

File: modules/videoio/include/opencv2/videoio.hpp [hunk 5]
```diff
@@ -791,10 +791,10 @@ class CV_EXPORTS_W VideoCapture
      *
      * methods raise exceptions if not successful instead of returning an error code
      */
-    CV_WRAP void setExceptionMode(bool enable) { throwOnFail = enable; }
+    CV_WRAP void setExceptionMode(bool enable) CV_NOEXCEPT { throwOnFail = enable; }
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: is noexcept really important here? let's undo it

File: modules/videoio/include/opencv2/videoio.hpp [hunk 6]
```diff
@@ -791,10 +791,10 @@ class CV_EXPORTS_W VideoCapture
...
 
     /// query if exception mode is active
-    CV_WRAP bool getExceptionMode() { return throwOnFail; }
+    CV_WRAP bool getExceptionMode() const CV_NOEXCEPT { return throwOnFail; }
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: same

File: modules/videoio/include/opencv2/videoio.hpp [hunk 7]
```diff
@@ -820,7 +820,7 @@ class CV_EXPORTS_W VideoCapture
 protected:
     Ptr<CvCapture> cap;
     Ptr<IVideoCapture> icap;
-    bool throwOnFail;
+    bool throwOnFail { false };
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: it's initialized in the constructor. why initialize it here?

File: modules/videoio/include/opencv2/videoio.hpp [hunk 8]
```diff
@@ -872,21 +872,21 @@ class CV_EXPORTS_W VideoWriter
       (eg. FFMPEG FFV1, Huffman HFYU, Lagarith LAGS, etc...)
     - If FFMPEG is enabled, using `codec=0; fps=0;` you can create an uncompressed (raw) video file.
     */
-    CV_WRAP VideoWriter(const String& filename, int fourcc, double fps,
-                Size frameSize, bool isColor = true);
+    CV_WRAP VideoWriter(const String& filename, int fourcc, double fps, const Size& frameSize,
+                        bool isColor = true);
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: let's undo it. no enough reason to break compatibility

File: modules/videoio/include/opencv2/videoio.hpp [hunk 9]
```diff
@@ -872,21 +872,21 @@ class CV_EXPORTS_W VideoWriter
...
 
     /** @overload
     The `apiPreference` parameter allows to specify API backends to use. Can be used to enforce a specific reader implementation
     if multiple are available: e.g. cv::CAP_FFMPEG or cv::CAP_GSTREAMER.
      */
     CV_WRAP VideoWriter(const String& filename, int apiPreference, int fourcc, double fps,
-                Size frameSize, bool isColor = true);
+                        const Size& frameSize, bool isColor = true);
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: same

File: modules/videoio/include/opencv2/videoio.hpp [hunk 10]
```diff
@@ -872,21 +872,21 @@ class CV_EXPORTS_W VideoWriter
...
 
     /** @brief Default destructor
 
     The method first calls VideoWriter::release to close the already opened file.
     */
-    virtual ~VideoWriter();
+    ~VideoWriter();
```
Comment (vpisarev) [inline modules/videoio/include/opencv2/videoio.hpp]: please, undo

File: modules/videoio/src/cap.cpp [hunk 11]
```diff
@@ -50,21 +50,34 @@ static bool param_VIDEOIO_DEBUG = utils::getConfigurationParameterBool("OPENCV_V
...
+{
+    cvReleaseVideoWriter(&obj);
+}
 
 
-VideoCapture::VideoCapture() : throwOnFail(false)
-{}
+VideoCapture::VideoCapture() = default;
 
-VideoCapture::VideoCapture(const String& filename, int apiPreference) : throwOnFail(false)
```
Comment (vpisarev) [inline modules/videoio/src/cap.cpp]: let's restore it

**Comment by vpisarev** [comment]
@VadimLevin, if I remember correctly, the goal was to add some profiles to video writer to enhance functionality. Here I just see many changes and absolutely no steps towards the goal that we stated. I have no principal objections against the changes in .cpp, as long as it passes all the tests. But besides adding "explicit" to the constructor, please, undo all the changes in header files, because a questionably better style is too high price for the lost compatibility.

**Comment by VadimLevin** [comment]
@vpisarev Yes, I separated work into several independent steps and this was one of them.
`virtual` methods in this class may lead to unexpected behavior for users, those want to inherited their classes from `VideoWriter` and `VideoCapture` because `open` and `release` are called in constructor and destructor respectively. If inheritance is not supposed to be, so there is no reasons to leave virtual methods in these classes at all (I try to follow the rule [C.132: Don’t make a function virtual without reason][https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c132-dont-make-a-function-virtual-without-reason] from Cpp Core Guidelines)

**Comment by asmorkalov** [comment]
@vpisarev @alalek Please take a look again.
@VadimLevin Please update PR description after scope change.

**Comment by asmorkalov** [comment]
@vpisarev friendly reminder.

**Pull request merged**